File: guess.ml
let game (answer : int) : int =
  let count = ref 0 in 

  let rec play () =
    print_string "Guess a number: ";
    let guess = read_int () in
    count := !count + 1;
    match compare guess answer with
    | -1 ->
      print_endline "Too Low";
      play ()
    | 1 ->
      print_endline "Too High";
      play ()
    | _ (* 0 *) -> 
      print_endline "Correct";
      !count
  in play ()
        
let main = 
  Random.self_init ();
  let answer = Random.int 100 in
  let guesses = game answer in
  print_endline ("It look " ^ (string_of_int guesses) ^ " tries.")

File: guess.hs
-- Based on https://wiki.haskell.org/Simple_StateT_use
import Control.Monad.State
import System.Random

game :: Int -> IO Int
game answer =
  execStateT play 0
  where
    play :: StateT Int IO ()
    play = do
      lift $ putStrLn "Guess a number:"
      input <- lift getLine
      let guess = read input
      modify (+ 1)
      case compare guess answer of
        LT -> do
          lift $ putStrLn "Too low"
          play
        GT -> do
          lift $ putStrLn "Too high"
          play
        EQ -> lift $ putStrLn "Correct"

main = do
  answer <- randomRIO (1, 100)
  guesses <- game answer
  putStrLn $ "It took " ++ show guesses ++ " tries."